src/push-state-tree.js

```diff
@@ -56,6 +51,59 @@ const CHANGE = 'change';
 const MATCH = 'match';
 const OLD_MATCH = 'oldMatch';
 
+// Internal history keep tracking of all changes in the location history, it can be reset any
+let internalHistory;
```

Comment:
Probably I will be able to remove this, I have a new strategy, but is highly more complex them it is right now, so I didn't remove this yet.

The ideia is to make a un-mutable copy of the events rules tree, them when the router is triggered again, it compares the tree for changes, if there is a removal, it executes the leave event from removed rules, and the execute enter on added rules if they match.

By doing this way, I won't need to track the location at all, just need to have stored the latest executed location string, like a global variable for all router scopes.

Maybe add a TODO here and leave unfinished? I wish to do that in another PR since all tests are passing in the PR already (it's stable).

What do you think about this plan?